src/Adapter/Order/OrderProductUpdater.php

```diff
@@ -0,0 +1,545 @@
+ * that is bundled with this package in the file LICENSE.txt.
+ * It is also available through the world-wide-web at this URL:
+ * https://opensource.org/licenses/OSL-3.0
+ * If you did not receive a copy of the license and are unable to
+ * obtain it through the world-wide-web, please send an email
+ * to license@prestashop.com so we can send you a copy immediately.
+ *
+ * DISCLAIMER
+ *
+ * Do not edit or add to this file if you wish to upgrade PrestaShop to newer
+ * versions in the future. If you wish to customize PrestaShop for your
+ * needs please refer to https://www.prestashop.com for more information.
+ *
+ * @author    PrestaShop SA <contact@prestashop.com>
+ * @copyright 2007-2020 PrestaShop SA and Contributors
+ * @license   https://opensource.org/licenses/OSL-3.0 Open Software License (OSL 3.0)
+ * International Registered Trademark & Property of PrestaShop SA
+ */
+
+declare(strict_types=1);
+
+namespace PrestaShop\PrestaShop\Adapter\Order;
+
+use Address;
+use Cart;
+use CartRule;
+use Configuration;
+use Context;
+use Country;
+use Currency;
+use Customer;
+use Db;
+use Language;
+use Order;
+use OrderCarrier;
+use OrderCartRule;
+use OrderDetail;
+use Pack;
+use PrestaShop\PrestaShop\Adapter\ContextStateManager;
+use PrestaShop\PrestaShop\Adapter\Order\Refund\OrderProductRemover;
+use PrestaShop\PrestaShop\Adapter\StockManager;
+use PrestaShop\PrestaShop\Core\Domain\Order\Exception\OrderException;
+use PrestaShop\PrestaShop\Core\Localization\CLDR\ComputingPrecision;
+use Product;
+use StockAvailable;
+use StockManagerFactory;
+use StockMvt;
+use Tools;
+use Validate;
+use Warehouse;
+
+/**
+ * Increase or decrease quantity of an order's product.
+ * Recalculate cart rules, order's prices and shipping infos.
+ */
+class OrderProductUpdater
+{
+    /**
+     * @var OrderAmountUpdater
+     */
+    private $orderAmountUpdater;
+
+    /**
+     * @var ContextStateManager
+     */
+    private $contextStateManager;
+
+    /**
+     * @var OrderProductRemover
+     */
+    private $orderProductRemover;
+
+    public function __construct(
+        OrderAmountUpdater $orderAmountUpdater,
+        OrderProductRemover $orderProductRemover,
+        ContextStateManager $contextStateManager
+    ) {
+        $this->orderAmountUpdater = $orderAmountUpdater;
+        $this->orderProductRemover = $orderProductRemover;
+        $this->contextStateManager = $contextStateManager;
+    }
+
+    /**
+     * @param Order $order
+     * @param OrderDetail $orderDetail
+     * @param int $oldQuantity
+     * @param int $newQuantity
+     * @param bool|null $productDeletionMode
+     * @param bool|null $hasInvoice
+     *
+     * @return Order
+     *
+     * @throws OrderException
+     * @throws \PrestaShopDatabaseException
+     * @throws \PrestaShopException
+     */
+    public function update(
+        Order $order,
+        OrderDetail $orderDetail,
+        int $oldQuantity,
```

Comment:
I believe this parameter is useless since it's always equal to `$orderDetail->product_quantity` ?

Reply:
In UpdateProductInOrderHandler we need to update the orderDetail quantity AND keep the old value to update the cart and stock

Reply:
It's not used any more in `UpdateProductInOrderHandler` (only to pass it to this method), so it's not needed to ave this extra parameter

Reply:
My point is that we need in OrderProductUpdater to know what was the original value of the quantity. I can do the `set` in the service 
$oldQuantity = $orderDetail->product_quantity
$orderDetail->product_quantity = $newQuantity

Reply:
done